x11/xbomb/xbomb.6
.\" This file Copyright 1994-2008 Andrew M. Bishop.
.\" It may be distributed under the GNU Public License, version 2, or
.\" any higher version.  See section COPYING of the GNU Public license
.\" for conditions under which this file may be redistributed.
.\"
.\" XBOMB manual page
.\" By Andrew M. Bishop November 94 + January 95 + January 2008
.\"
.TH XBomb 1 "January 5, 2008"
.SH NAME
xbomb Version 2 \- a version of the popular minesweeper program with some novel features.
.SH SYNOPSIS
.B xbomb
.BR [\-hexagon | \-square | \-triangle]
.BR [\-1 | \-2 | \-3] 
.BR [\-hiscore]
.SH OPTIONS
.TP 
.BR \-hiscore
command line option displays the current hi-score table without starting the game.
.TP 
.BR \-hexagon \ | \ \-square \ | \ \-triangle
Selects the type of grid to use when starting the game.
.TP 
.BR \-1 \ | \ \-2 \ | \ \-3
Selects the level to use when starting the game.
.TP
The standard X window options also apply to this program.
.SH DESCRIPTION
A game of logical deduction against the clock. The purpose of the game is to find the location of the hidden bombs in a grid.
Each of the locations in the grid is initially covered, by uncovering a locations, a number or a bomb will be revealed underneath.
If a bomb is revealed then the game is finished.
A number indicates the number of adjacent locations that contain bombs.
Locations that have not been uncovered but are thought to contain bombs can be marked as such so that they can be identified.

There is a choice of hexagonal, square or triangular units that make up the grid, the level of difficulty increasing as the number of sides decreases.
In all cases the grids are of the same size and contain the same number of bombs for the same levels.

There are three levels to the game:
.TP
.BR Easy
8 by 8 grid with 10 bombs.
.TP
.BR Medium
16 by 16 grid with 40 bombs.
.TP
.BR Difficult
30 by 16 grid with 99 bombs.
.LP
A high score table is maintained for each of the levels for each of the grid types,
storing the username, the time to complete the puzzle and the date that it was finished.
.SH MOUSE OPERATION
The game is played with the mouse, with all three buttons having different actions.
.TP
.BR Left \ Button
Uncover the current location if it is hidden or uncover all surrounding
locations marked as not containing bombs if the square is already uncovered. The
first selection made in a game is always a 'safe' square.
.TP
.BR Middle \ Button
Uncover all surrounding locations that are not marked as containing a bomb.
.TP
.BR Right \ Button
Mark the current grid location as having a hidden bomb (toggle).
.SH KEYSTROKES
There are a number of keystroke shortcuts for the game:
.TP
.BR 1 \ 2 \ 3
Select level 1, 2 or 3 for the next game.
.TP
.BR s
Start a new game.
.TP
.BR q
Quit the program.
.TP
.BR h
Print the high-scores.
.TP
.BR H \ or \ Control-H
Use the grid of hexagons for the next game.
.TP
.BR S \ or \ Control-S
Use the grid of squares for the next game.
.TP
.BR T \ or \ Control-T
Use the grid of triangles for the next game.
.TP
.SH X RESOURCES
The name of the class used is XBomb, the widget hierarchy used is (Widget-Class:name):

 	XBomb
 		Form:form
 			Command:start
 			MenuButton:menu
 				SimpleMenu:levelmenushell
 					smeBSBObject:Easy
 					smeBSBObject:Medium
 					smeBSBObject:Difficult
 			MenuButton:menu
 				SimpleMenu:typemenushell
 					smeBSBObject:Hexagon
 					smeBSBObject:Square
 					smeBSBObject:Triangle
 			Command:hiscore
 			Command:quit
 			Core:grid
 			Command:clock
 			Command:uxb

.LP
The extra resources that are available for the grid are:
.TP
.BR grid.bomb
The colour of the bombs \- default black.
.TP
.BR grid.bombreal
The colour of the real bombs \- default red.
.TP
.BR grid.bombmark
The colour of the bombs that are marked by the user \- default orange.
.TP
.BR grid.number
The colour of the number '1' when drawn in the grid \- default navy blue.
.TP
.BR grid.number2
The colour of the number '2' when drawn in the grid \- default dark green.
.TP
.BR grid.number3
The colour of the number '3' when drawn in the grid \- default dark red.
The other numbers up to 12 also have their own resources.
.TP
.BR grid.hidden
The colour of the hidden squares \- default grey50.
.TP
.BR grid.correct
The colour of the squares that were correctly guessed \- default green
.TP
.BR grid.font
The font that is used for the number in the grid \- default lucidasans-24
.SH BUGS
None known.
.SH AUTHOR
Andrew M. Bishop.


x11/xbomb/xbomb.c
/***************************************
  XBomb - 'Minesweeper' game - Version 2.2b.

  Main function of program - excludes all window interface parts.
  ******************/ /******************
  Written by Andrew M. Bishop

  This file Copyright 1994-2014 Andrew M. Bishop
  It may be distributed under the GNU Public License, version 2, or
  any higher version.  See section COPYING of the GNU Public license
  for conditions under which this file may be redistributed.
  ***************************************/


#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>
#include <ctype.h>

#if defined(__sun__) && !defined(__svr4__)
int    printf(const char*, ... );
int    tolower (int c);
time_t time(time_t *t);
int    rand(void);
#endif

#include "xbomb.h"

/*+ The names of the different game +*/
char *levels[NLEVELS]={"Easy","Medium","Difficult"},      /*+ levels (difficulty). +*/
     *types [NTYPES] ={"Hexagons","Squares","Triangles"}; /*+ types (grid shapes). +*/

/*+ The size of the grids +*/
int widths [NLEVELS]={ 8 ,16 ,30},  /*+ width in tiles. +*/
    heights[NLEVELS]={ 8 ,16 ,16},  /*+ height in tiles. +*/
    nbombs [NLEVELS]={10 ,40 ,99};  /*+ number of bombs. +*/

/*+ The current game status +*/
static int status=GAME_WAIT;

/*+ The current information about the grid +*/
int grid_width,                 /*+ width. +*/
    grid_height,                /*+ height. +*/
    grid_bombs,                 /*+ number of bombs. +*/
    grid_level,                 /*+ level. +*/
    grid_type;                  /*+ type. +*/

/*+ The status of the tiles in the grid. +*/
static unsigned char state[MAX_SIZE][MAX_SIZE];

/*+ The status of the game, +*/
static int n_unseen,            /*+ the number of tiles unseen. +*/
           n_think,             /*+ the number of flags put down. +*/
           ticks;               /*+ The number of milliseconds of time passed. +*/


static int count_adjacent(int x,int y,int flag);


/*++++++++++++++++++++++++++++++++++++++
  The main program.

  int argc The command line parameters.

  char** argv The command line parameters.
  ++++++++++++++++++++++++++++++++++++++*/

int main(int argc,char** argv)
{
 int level=GAME_EASY,type=GAME_SQUARE;  /* Default values. */
 int xstatus=0;
 long t;

 printf("\n"
        "XBomb Version 2.2b\n"
        "\n"
        "(c) Andrew M. Bishop 1994-2013 [http://www.gedanken.org.uk/software/xbomb/]\n"
        "\n");

 if(argc>1)
   {
    int i,j;

    for(j=0;j<NLEVELS;j++)
      {
       char levstr[3];
       sprintf(levstr,"-%1d",j+1);
       for(i=1;i<argc;i++)
          if(!strcmp(levstr,argv[i]))
             level=GAME_LEVEL+j;
      }

    for(j=0;j<NTYPES;j++)
      {
       for(i=1;i<argc;i++)
          if(argv[i][0]=='-' && !strncasecmp(types[j],argv[i]+1,strlen(argv[i])-1))
             type=GAME_TYPE+j;
      }

    for(i=1;i<argc;i++)
       if(strlen(argv[i])>2 && !strncmp("-hiscore",argv[i],strlen(argv[i])))
         {
          grid_type=type;
          PrintHighScores();
          exit(0);
         }
   }

 t=time(&t);

 srand((unsigned int)t);

 InitialiseX(&argc,argv);

 StartGame(level,type);
 ShowHighScores();

 for(;;)
   {
    xstatus=ProcessXEvents();

    if(xstatus==GAME_QUIT)
       break;

    if(xstatus>=GAME_LEVEL && xstatus<(GAME_LEVEL+NLEVELS) && xstatus!=level)
      {
       level=xstatus;
       xstatus=GAME_START;
      }

    if(xstatus>=GAME_TYPE && xstatus<(GAME_TYPE+NTYPES) && xstatus!=type)
      {
       type=xstatus;
       xstatus=GAME_START;
      }

    if(xstatus==GAME_START)
      {
       if(status==GAME_CONTINUES)
          StopGame();
       StartGame(level,type);
       ShowHighScores();
      }

    if(status==GAME_WON)
      {
       int x,y;
       for(x=0;x<grid_width;x++)
          for(y=0;y<grid_height;y++)
             state[x][y]|=CORRECT;
       DrawGrid();
       StopGame();
       AddHighScore(ticks);
       ShowHighScores();
      }
    else if(status==GAME_LOST)
      {
       StopGame();
       ShowHighScores();
      }
   } 

 FinishUpX();

 return(0);
}


/*++++++++++++++++++++++++++++++++++++++
  Start a new game.

  int level The level of the game to play.

  int type The type of game to play.
  ++++++++++++++++++++++++++++++++++++++*/

void StartGame(int level,int type)
{
 int x,y;

 grid_type=type;
 grid_level=level;
 grid_width=widths[level-GAME_LEVEL];
 grid_height=heights[level-GAME_LEVEL];
 grid_bombs=nbombs[level-GAME_LEVEL];

 n_unseen=grid_width*grid_height;
 n_think=0;

 status=GAME_READY;

 for(x=0;x<grid_width;x++)
    for(y=0;y<grid_height;y++)
       state[x][y]=UNSEEN;

 ScaleWindow();
 DrawGrid();
 StartClock(0);
 SetUXB(grid_bombs);
}


/*++++++++++++++++++++++++++++++++++++++
  Hides the bombs in the grid.

  int xs The location that has no neighbouring bombs.

  int ys The location that has no neighbouring bombs.
  ++++++++++++++++++++++++++++++++++++++*/

void HideBombs(int xs,int ys)
{
 int i,x,y;

 state[xs][ys]+=CORRECT;

 for(i=0;i<grid_bombs;i++)
   {
    do{
       x=(rand()>>8)%grid_width;
       y=(rand()>>8)%grid_height;
      }
    while(state[x][y]!=UNSEEN || count_adjacent(x,y,CORRECT));

    state[x][y]+=ACTUAL_BOMB;
   }

 state[xs][ys]-=CORRECT;

 for(x=0;x<grid_width;x++)
    for(y=0;y<grid_height;y++)
       if(!(state[x][y]&ACTUAL_BOMB))
          state[x][y]+=count_adjacent(x,y,ACTUAL_BOMB);

 StartClock(1);
}


/*++++++++++++++++++++++++++++++++++++++
  Stop the game.
  ++++++++++++++++++++++++++++++++++++++*/

void StopGame(void)
{
 status=GAME_WAIT;
 ticks=StopClock();
}


/*++++++++++++++++++++++++++++++++++++++
  Draws the grid at the start of the game or for expose events.
  ++++++++++++++++++++++++++++++++++++++*/

void DrawGrid(void)
{
 int x,y;

 for(x=0;x<grid_width;x++)
    for(y=0;y<grid_height;y++)
       DrawSquare(x,y,state[x][y]);
}


/*++++++++++++++++++++++++++++++++++++++
  Select a single square and clear it or explode the bomb.

  int x The location of the square.

  int y The location of the square.
  ++++++++++++++++++++++++++++++++++++++*/

void SelectSquare(int x,int y)
{
 if(!(state[x][y]&UNSEEN) || state[x][y]&THINK_BOMB || status==GAME_WAIT)
    return;

 if(status==GAME_READY)
   {
    HideBombs(x,y);
    status=GAME_CONTINUES;
   }

 RemoveEmpties(x,y);

 if(state[x][y]&ACTUAL_BOMB)
   {
    int x,y;

    for(x=0;x<grid_width;x++)
       for(y=0;y<grid_height;y++)
          if(state[x][y]&UNSEEN)
            {
             if(!(state[x][y]&THINK_BOMB))
                state[x][y]&=~UNSEEN;
             if(state[x][y]&ACTUAL_BOMB)
                state[x][y]|=CORRECT;
            }
          else if(!(state[x][y]&ACTUAL_BOMB))
             state[x][y]|=CORRECT;

    status=GAME_LOST;
    DrawGrid();
   }
 else
    if(n_think==n_unseen)
       status=GAME_WON;
}


/*++++++++++++++++++++++++++++++++++++++
  Select a single square and clear it, explode the bomb or clear the adjacent.

  int x The location of the square.

  int y The location of the square.
  ++++++++++++++++++++++++++++++++++++++*/

void SelectSquareOrAdjacent(int x,int y)
{
 if(state[x][y]&UNSEEN)
    SelectSquare(x,y);
 else /* if(state[x][y]&EMPTY) */
    SelectAdjacent(x,y);
}


/*++++++++++++++++++++++++++++++++++++++
  Select all of the squares adjacent

  int x The location of the square.

  int y The location of the square.
  ++++++++++++++++++++++++++++++++++++++*/

void SelectAdjacent(int x,int y)
{
 int dx,dy,n,dd=(grid_type==GAME_TRIANGLE?2:1);

 if(state[x][y]&UNSEEN || status==GAME_WAIT)
    return;

 n=count_adjacent(x,y,THINK_BOMB);

 if(state[x][y]!=n)
    return;

 for(dx=-dd;dx<=dd;dx++)
   {
    if((x+dx)<0 || (x+dx)>=grid_width)
       continue;

    for(dy=-1;dy<=1;dy++)
      {
       if((y+dy)<0 || (y+dy)>=grid_height)
          continue;
       if(dx==0 && dy==0)
          continue;

       if(grid_type==GAME_HEXAGON && dy!=0 && dx==(1-2*(y%2)))
          continue;
       if(grid_type==GAME_TRIANGLE && dy==(2*((x+y)%2)-1) && (dx==-2 || dx==2))
          continue;

       if(!(state[x+dx][y+dy]&THINK_BOMB))
          SelectSquare(x+dx,y+dy);
      }
   }
}


/*++++++++++++++++++++++++++++++++++++++
  Marks as a bomb the square.

  int x The location of the square.

  int y The location of the square.
  ++++++++++++++++++++++++++++++++++++++*/

void MarkBomb(int x,int y)
{
 if(!(state[x][y]&UNSEEN) || status==GAME_WAIT || status==GAME_READY)
    return;

 if(n_think==grid_bombs && !(state[x][y]&THINK_BOMB))
    return;

 state[x][y]^=THINK_BOMB;

 if(state[x][y]&THINK_BOMB)
    n_think++;
 else
    n_think--;

 SetUXB(grid_bombs-n_think);

 DrawSquare(x,y,state[x][y]);

 if(n_think==n_unseen)
    status=GAME_WON;
}


/*++++++++++++++++++++++++++++++++++++++
  Removes the empty squares around here

  int x Start from this position.

  int y Start from this position.
  ++++++++++++++++++++++++++++++++++++++*/

void RemoveEmpties(int x, int y)
{
 int dx,dy,dd=(grid_type==GAME_TRIANGLE?2:1);

 if(!(state[x][y]&UNSEEN) || state[x][y]&THINK_BOMB)
    return;

 state[x][y]&=~UNSEEN;
 n_unseen--;
 DrawSquare(x,y,state[x][y]);

 if(state[x][y]!=EMPTY)
    return;

 for(dx=-dd;dx<=dd;dx++)
   {
    if((x+dx)<0 || (x+dx)>=grid_width)
       continue;

    for(dy=-1;dy<=1;dy++)
      {
       if((y+dy)<0 || (y+dy)>=grid_height)
          continue;
       if(dx==0 && dy==0)
          continue;

       if(grid_type==GAME_HEXAGON && dy!=0 && dx==(1-2*(y%2)))
          continue;
       if(grid_type==GAME_TRIANGLE && dy==(2*((x+y)%2)-1) && (dx==-2 || dx==2))
          continue;

       RemoveEmpties(x+dx,y+dy);
      }
   }
}


/*++++++++++++++++++++++++++++++++++++++
  Count the number of adjacent squares that are bombs.

  int count_adjacent Returns the number of adjacent tiles with the specified flag.

  int x The position to start from.

  int y The position to start from.

  int flag The type of thing to look for.
  ++++++++++++++++++++++++++++++++++++++*/

static int count_adjacent(int x,int y,int flag)
{
 int n=0,dx,dy,dd=(grid_type==GAME_TRIANGLE?2:1);

 for(dx=-dd;dx<=dd;dx++)
   {
    if((x+dx)<0 || (x+dx)>=grid_width)
       continue;

    for(dy=-1;dy<=1;dy++)
      {
       if((y+dy)<0 || (y+dy)>=grid_height)
          continue;

       if(grid_type==GAME_HEXAGON && dy!=0 && dx==(1-2*(y%2)))
          continue;
       if(grid_type==GAME_TRIANGLE && dy==(2*((x+y)%2)-1) && (dx==-2 || dx==2))
          continue;

       if(state[x+dx][y+dy]&flag)
          n++;
      }
   }

 return(n);
}
